File: dotnet/DesignPatterns/BehavioralDesignPatterns/MementoDesignPattern/MementoDesignPattern.Example1/DocumentHistory.cs
namespace MementoDesignPattern.Example1;

/// <summary>
/// Caretaker: Manages the history of snapshots for a single TextEditor document.
/// It holds two stacks — one for undo operations and one for redo operations —
/// and orchestrates saving (Backup), reverting (Undo), and re-applying (Redo) states.
/// Crucially, it never reads or modifies the contents of a DocumentSnapshot;
/// it only stores and retrieves them, keeping the Memento's encapsulation intact.
/// </summary>
public class DocumentHistory 
{
    private readonly TextEditor _textEditor;

    // Undo / Redo stacks
    private readonly Stack<DocumentSnapshot> _undoStack = new();
    private readonly Stack<DocumentSnapshot> _redoStack = new();

    public DocumentHistory(TextEditor textEditor)
    {
        _textEditor = textEditor;
    }

    public void Backup(string snapshotName = "Auto-Save")
    {
        _undoStack.Push(_textEditor.CreateSnapshot(snapshotName));
        _redoStack.Clear();
    }

    public void Undo()
    {
        if(_undoStack.Count == 0)
        {
            Console.WriteLine("No more undo steps available.");
            return;
        }

        // Save current state to redo stack before undoing
        _redoStack.Push(_textEditor.CreateSnapshot("Before Undo"));

        // Restore previous state from undo stack
        var previousSnapshot = _undoStack.Pop();
        _textEditor.RestoreFromSnapshot(previousSnapshot);
    }

    public void Redo()
    {
        if(_redoStack.Count == 0)
        {
            Console.WriteLine("No more redo steps available.");
            return;
        }

        _undoStack.Push(_textEditor.CreateSnapshot("Before Redo"));

        // Restore next state from redo stack
        var nextSnapshot = _redoStack.Pop();
        _textEditor.RestoreFromSnapshot(nextSnapshot);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/MementoDesignPattern/MementoDesignPattern.Example1/src/DocumentHistory.ts
import { TextEditor } from './TextEditor';
import { DocumentSnapshot } from './DocumentSnapshot';

export class DocumentHistory {
  private readonly _textEditor: TextEditor;
  private readonly _undoStack: DocumentSnapshot[] = [];
  private readonly _redoStack: DocumentSnapshot[] = [];

  constructor(textEditor: TextEditor) {
    this._textEditor = textEditor;
  }

  backup(snapshotName: string = 'Auto-Save'): void {
    this._undoStack.push(this._textEditor.createSnapshot(snapshotName));
    this._redoStack.length = 0;
  }

  undo(): void {
    if (this._undoStack.length === 0) {
      console.log('No more undo steps available.');
      return;
    }

    this._redoStack.push(this._textEditor.createSnapshot('Before Undo'));
    const previousSnapshot = this._undoStack.pop()!;
    this._textEditor.restoreFromSnapshot(previousSnapshot);
  }

  redo(): void {
    if (this._redoStack.length === 0) {
      console.log('No more redo steps available.');
      return;
    }

    this._undoStack.push(this._textEditor.createSnapshot('Before Redo'));
    const nextSnapshot = this._redoStack.pop()!;
    this._textEditor.restoreFromSnapshot(nextSnapshot);
  }
}
